### PR TITLE
Editor: Refactor Post Author component

### PR DESCRIPTION
**Pull request by Mamaduka**
## Description
Refactors Editor's post author component to use `getUser` and `getUsers` selectors. 

The component previously was using temporary `getAuthors` and `__unstableGetAuthor` to display author selection for non-admin users. Temp selectors are no longer required as now we can specify context when requesting data records.

We can now deprecated/remove those temporary selectors.

### Query Improvements
* Selectors are now using shared query arguments objects—this way, Editor is only making _one request when rendering the component instead of four._
* Request only fields we need for rendering (`id, name`). The change _reduces transferred data size by roughly 50%._

## How has this been tested?
Test creating/editing post.

* One user - the post author component isn't visible
* 25 or fewer authors - Select dropdown is rendered.
* 25 and more authors - Combobox is rendered, allows searching users.

Test with non-admin users. Again, the results should be the same.

### Helpful WP CLI Commands

Generate 15 editors: `wp user generate --role=editor --count=15`
Generate 50 editors: `wp user generate --role=editor --count=50`
Cleanup: `wp user delete $(wp user list --role=editor --field=ID)`

## Screenshots 
| HTTP Requests Before  | HTTP Requests After |
| ------------- | ------------- |
| post-author-queries-before  | post-author-queries-after  |

## Types of changes
Refactoring

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/editor/src/components/post-author/check.js [hunk 1]
```diff
@@ -36,14 +36,14 @@ export function PostAuthorCheck( {
...
 	withSelect( ( select ) => {
 		const post = select( editorStore ).getCurrentPost();
+		const authors = select( coreStore ).getUsers( AUTHORS_QUERY );
 		return {
 			hasAssignAuthorAction: get(
 				post,
 				[ '_links', 'wp:action-assign-author' ],
 				false
 			),
-			postType: select( editorStore ).getCurrentPostType(),
```
Comment (Mamaduka) [inline packages/editor/src/components/post-author/check.js:44]: We weren't using this prop in `PostAuthorCheck.`

File: packages/editor/src/components/post-author/combobox.js [hunk 2]
```diff
@@ -57,22 +61,17 @@ function PostAuthorCombobox() {
...
 	}, [ authors, postAuthor ] );
 
-	// Initializes the post author properly
-	// Also ensures external changes are reflected.
-	useEffect( () => {
-		if ( postAuthor ) {
-			setFieldValue( postAuthor.name );
-		}
-	}, [ postAuthor ] );
-
```
Comment (Mamaduka) [inline packages/editor/src/components/post-author/combobox.js:75]: I couldn't track down why this effect was added precisely, but I don't think it's needed anymore.

It also initializes an unnecessary search query on the mount.

File: packages/editor/src/components/post-author/constants.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export const AUTHORS_QUERY = {
+	who: 'authors',
+	per_page: 50,
```
Comment (Mamaduka) [inline packages/editor/src/components/post-author/constants.js:3]: I changed the "per_page" argument from 100 to 50. However, I don't think this number makes a big difference when used with `ComboboxControl.` It just needs to be higher than 25 and reasonable :)

Happy to revert it if needed.

File: packages/editor/src/components/post-author/combobox.js [hunk 4]
```diff
@@ -46,7 +50,7 @@ function PostAuthorCombobox() {
 		const fetchedAuthors = ( authors ?? [] ).map( ( author ) => {
 			return {
 				value: author.id,
-				label: author.name,
+				label: decodeEntities( author.name ),
```
Comment (Mamaduka) [inline packages/editor/src/components/post-author/combobox.js:53]: Added missing entity decoding. We were already doing this for select control.

**Comment by Mamaduka** [comment]
@youknowriad, this isn't necessarily related to this PR, but I'm just curious. Is it okay to perform `map` and `filter` actions inside selector callbacks?

https://github.com/WordPress/gutenberg/blob/c6ba02bba3a4c89249f45b0371b0b1c8e070363f/packages/editor/src/components/post-author/select.js#L23-L26

File: packages/editor/src/components/post-author/check.js [hunk 5]
```diff
@@ -35,14 +36,14 @@ export function PostAuthorCheck( {
 export default compose( [
```
Comment (youknowriad) [inline packages/editor/src/components/post-author/check.js]: Nit: Could easily be transformed to hooks now.

File: packages/editor/src/components/post-author/constants.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+	_fields: 'id,name',
+	context: 'view',
```
Comment (youknowriad) [inline packages/editor/src/components/post-author/constants.js]: Maybe add a comment explaining that the context is "view" in order to allow non-admin roles perform the request.

File: packages/editor/src/components/post-author/index.js [hunk 7]
```diff
@@ -9,13 +9,14 @@ import { store as coreStore } from '@wordpress/core-data';
...
 import PostAuthorCombobox from './combobox';
 import PostAuthorSelect from './select';
+import { AUTHORS_QUERY } from './constants';
 
 const minimumUsersForCombobox = 25;
 
 function PostAuthor() {
 	const showCombobox = useSelect( ( select ) => {
-		// Not using `getUsers()` because it requires `list_users` capability.
-		const authors = select( coreStore ).getAuthors();
```
Comment (youknowriad) [inline packages/editor/src/components/post-author/index.js:18]: Should we deprecate this selector and remove the unstable one?

**Review by youknowriad — COMMENTED** [review]
@Mamaduka mapping/filtering inside useSelect/withSelect should definitely be avoided as it will trigger too much rerendering. Favor `useMemo` outside the `useSelect` for these situations.

Awesome work on this PR btw

File: packages/editor/src/components/post-author/index.js (hunk 7 above)
Comment (Mamaduka) [reply packages/editor/src/components/post-author/index.js:18]: Yes, that's the plan for the next PR :)

File: packages/editor/src/components/post-author/check.js (hunk 5 above)
Comment (Mamaduka) [reply packages/editor/src/components/post-author/check.js]: I guess this means we should remove the test for `PostAuthorCheck`. The component won't be really unit-testable after this change.

Comment (Mamaduka) [reply packages/editor/src/components/post-author/check.js]: Never mind, I found examples of how to update the tests.

Comment (youknowriad) [reply packages/editor/src/components/post-author/check.js]: To be honest, these tests are not the best we have either :)

**Review by youknowriad — APPROVED** [review]
LGTM 👍

**Comment by Mamaduka** [comment]
Thanks for the review, @youknowriad.

**Pull request merged**